### PR TITLE
Check for float values for min/max values to ax{v,h}line

### LINKED ISSUE #12198: axvline incorrectly tries to handle unitized ymin, ymax
After
```
line = plt.axvline(
    x=.5, ymin=matplotlib.dates.num2date(1), ymax=matplotlib.dates.num2date(1.05))
line.set_clip_on(False)
```
and panning the axes around, one sees that matplotlib (3.0) converted ymin and ymax to the scalar values 1 and 1.05, which is not correct: ymin and ymax are values relative to the axes bounds and thus intrinsically unitless; passing unitized data to ymin or ymax should result in an error.

The same point likely applies to axhline, axvspan, axhspan.

### PR DESCRIPTION
**Pull request by dstansby**
Fixes #12198. This now errors if a non-float value is provided for xmin/xmax or ymin/ymax to ax{v,h}{line,span}. This should get an API change or a what's new, but I'm not sure which, could someone advise?

### CODE REVIEW

**Comment by tacaswell** [comment]
I don't think checking for the ability to cast to float is enough, we need to make sure that the data will not select a converter.

**Comment by dstansby** [comment]
I found `matplotlib.units._is_natively_supported`, which I think does the correct job here.

**Comment by jklymak** [comment]
On master 

```python
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
line = plt.axvline(
    x=.5, ymin=mdates.num2date(1), ymax=mdates.num2date(1.05))
line.set_clip_on(False)

plt.show()
```
already errors out:

```
Traceback (most recent call last):
  File "./testAXvline.py", line 3, in <module>
    line = plt.axvline(
  File "/Users/jklymak/matplotlib/lib/matplotlib/pyplot.py", line 2464, in axvline
    return gca().axvline(x=x, ymin=ymin, ymax=ymax, **kwargs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_axes.py", line 917, in axvline
    self.add_line(l)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_base.py", line 1977, in add_line
    self._update_line_limits(line)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_base.py", line 1999, in _update_line_limits
    path = line.get_path()
  File "/Users/jklymak/matplotlib/lib/matplotlib/lines.py", line 1011, in get_path
    self.recache()
  File "/Users/jklymak/matplotlib/lib/matplotlib/lines.py", line 658, in recache
    y = _to_unmasked_float_array(yconv).ravel()
  File "/Users/jklymak/matplotlib/lib/matplotlib/cbook/__init__.py", line 1298, in _to_unmasked_float_array
    return np.asarray(x, float)
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.8/site-packages/numpy/core/_asarray.py", line 85, in asarray
    return array(a, dtype, copy=False, order=order)
TypeError: float() argument must be a string or a number, not 'datetime.datetime'
(matplotlibdev)
```

So is the point of this PR to just make the error more explicit?   If so, I don't  think it needs an API change or note, because I guess that has already occurred.

**Comment by QuLogic** [comment]
That specific example broke in a793bde0571709597d7434c4bdbf6e3ad5cd183b, which is a date-specific change. Is it still true that other unit-ful values are currently broken?

**Comment by dstansby** [comment]
Yes, I think it might just be a better error now, with the following example
```python
import matplotlib.pyplot as plt
import astropy.units as u
from astropy.visualization import quantity_support
quantity_support()

fig, ax = plt.subplots()
ax.scatter(1 * u.m, 1 * u.s)
ax.axhline(1 * u.s, xmin=0.1 * u.m)
plt.show()
```

The error before was
```python
Traceback (most recent call last):
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/axis.py", line 1524, in convert_units
    ret = self.converter.convert(x, self.units, self)
  File "/Users/dstansby/github/astropy/astropy/visualization/units.py", line 102, in convert
    return [v.to_value(unit) for v in val]
  File "/Users/dstansby/github/astropy/astropy/visualization/units.py", line 102, in <listcomp>
    return [v.to_value(unit) for v in val]
AttributeError: 'int' object has no attribute 'to_value'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 8, in <module>
    ax.axhline(1 * u.s, xmin=0.1 * u.m)
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/axes/_axes.py", line 847, in axhline
    self.add_line(l)
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/axes/_base.py", line 1977, in add_line
    self._update_line_limits(line)
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/axes/_base.py", line 1999, in _update_line_limits
    path = line.get_path()
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/lines.py", line 1011, in get_path
    self.recache()
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/lines.py", line 652, in recache
    xconv = self.convert_xunits(self._xorig)
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/artist.py", line 175, in convert_xunits
    return ax.xaxis.convert_units(x)
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/axis.py", line 1526, in convert_units
    raise munits.ConversionError('Failed to convert value(s) to axis '
matplotlib.units.ConversionError: Failed to convert value(s) to axis units: [<Quantity 0.1 m>, 1]
```

And the error with this PR is
```python
Traceback (most recent call last):
  File "test.py", line 8, in <module>
    ax.axhline(1 * u.s, xmin=0.1 * u.m)
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/axes/_axes.py", line 834, in axhline
    self._check_no_units([xmin, xmax], ['xmin', 'xmax'])
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/axes/_axes.py", line 926, in _check_no_units
    raise ValueError(f"{name} must be a single scalar value, "
ValueError: xmin must be a single scalar value, but got 0.1 m
```

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -830,10 +831,10 @@ def axhline(self, y=0, xmin=0, xmax=1, **kwargs):
 
             >>> axhline(y=.5, xmin=0.25, xmax=0.75)
         """
+        self._check_no_units([xmin, xmax], ['xmin', 'xmax'])
         if "transform" in kwargs:
-            raise ValueError(
-                "'transform' is not allowed as a kwarg;"
-                + "axhline generates its own transform.")
+            raise ValueError("'transform' is not allowed as a kwarg; "
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py:836]: This should be spelled out:
```suggestion
            raise ValueError("'transform' is not allowed as a keyword argument; "
```

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -899,11 +900,10 @@ def axvline(self, x=0, ymin=0, ymax=1, **kwargs):
 
             >>> axvline(x=.5, ymin=0.25, ymax=0.75)
         """
-
+        self._check_no_units([ymin, ymax], ['ymin', 'ymax'])
         if "transform" in kwargs:
-            raise ValueError(
-                "'transform' is not allowed as a kwarg;"
-                + "axvline generates its own transform.")
+            raise ValueError("'transform' is not allowed as a kwarg; "
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py:905]:
```suggestion
            raise ValueError("'transform' is not allowed as a keyword argument; "
```

**Comment by QuLogic** [comment]
Hmm, those exception lines are too long; I wonder why `flake8` is not complaining.

**Pull request merged**